src/cargo/core/resolver/mod.rs

```diff
@@ -298,109 +244,213 @@ fn activate(mut cx: Box<Context>,
+            uses_default_features: dep.uses_default_features(),
+            target_platform: platform.as_ref().map(|p| p as &str),
+        };
 
-    let prev_active = cx.prev_active(dep);
-    trace!("{}[{}]>{} {} candidates", parent.name(), cur, dep.name(),
-           candidates.len());
-    trace!("{}[{}]>{} {} prev activations", parent.name(), cur,
-           dep.name(), prev_active.len());
-
-    // Filter the set of candidates based on the previously activated
-    // versions for this dependency. We can actually use a version if it
-    // precisely matches an activated version or if it is otherwise
-    // incompatible with all other activated versions. Note that we define
-    // "compatible" here in terms of the semver sense where if the left-most
-    // nonzero digit is the same they're considered compatible.
-    let my_candidates = candidates.iter().filter(|&b| {
-        prev_active.iter().any(|a| a == b) ||
-            prev_active.iter().all(|a| {
-                !compatible(a.version(), b.version())
-            })
-    });
+        let prev_active: Vec<Rc<Summary>> = cx.prev_active(&dep).to_vec();
+        trace!("{}[{}]>{} {} candidates", parent.name(), cur, dep.name(),
+               candidates.len());
+        trace!("{}[{}]>{} {} prev activations", parent.name(), cur,
+               dep.name(), prev_active.len());
+
+        // Filter the set of candidates based on the previously activated
+        // versions for this dependency. We can actually use a version if it
+        // precisely matches an activated version or if it is otherwise
+        // incompatible with all other activated versions. Note that we define
+        // "compatible" here in terms of the semver sense where if the left-most
+        // nonzero digit is the same they're considered compatible.
+        let my_candidates: Vec<Rc<Summary>> = candidates.iter().filter(|&b| {
+            prev_active.iter().any(|a| a == b) ||
+                prev_active.iter().all(|a| {
+                    !compatible(a.version(), b.version())
+                })
+        }).cloned().collect();
+
+        // Alright, for each candidate that's gotten this far, it meets the
+        // following requirements:
+        //
+        // 1. The version matches the dependency requirement listed for this
+        //    package
+        // 2. There are no activated versions for this package which are
+        //    semver-compatible, or there's an activated version which is
+        //    precisely equal to `candidate`.
+        //
+        // This means that we're going to attempt to activate each candidate in
+        // turn. We could possibly fail to activate each candidate, so we try
+        // each one in turn.
+        let mut remaining_candidates = RcVecIter::new(my_candidates);
+        let maybe_candidate =
+            remaining_candidates.next().map(|(_, candidate)| candidate.clone());
+        let candidate: Rc<Summary> = match maybe_candidate {
```

Comment:
When I first tried reading this code, I found it very hard to keep the expected types of variables in my head, so as a courtesy to the next person trying to come up to speed on this code, I left in the annotations that helped me understand it. Obviously you're the maintainer so I'll take them back out if you want, but I think it could use more than the bare minimum.